src/python/pants/backend/python/rules/black.py

```diff
@@ -0,0 +1,149 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Optional, Tuple
+
+from pants.backend.python.rules.pex import (
+  CreatePex,
+  Pex,
+  PexInterpreterConstraints,
+  PexRequirements,
+)
+from pants.backend.python.subsystems.black import Black
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.backend.python.subsystems.subprocess_environment import SubprocessEncodingEnvironment
+from pants.backend.python.targets.formattable_python_target import FormattablePythonTarget
+from pants.engine.fs import Digest, DirectoriesToMerge, PathGlobs, Snapshot
+from pants.engine.isolated_process import (
+  ExecuteProcessRequest,
+  ExecuteProcessResult,
+  FallibleExecuteProcessResult,
+)
+from pants.engine.rules import optionable_rule, rule
+from pants.engine.selectors import Get
+from pants.rules.core.fmt import FmtResult
+from pants.rules.core.lint import LintResult
+
+
+@dataclass(frozen=True)
+class BlackSetup:
+  """This abstraction is used to deduplicate the implementations for the `fmt` and `lint` rules,
+  which only differ in whether or not to append `--check` to the Black CLI args."""
+  config_path: Optional[Path]
+  resolved_requirements_pex: Pex
+  merged_input_files: Digest
+
+  def generate_pex_arg_list(self, *, files: Tuple[str, ...], check_only: bool) -> Tuple[str, ...]:
+    pex_args = []
+    if check_only:
+      pex_args.append("--check")
+    if self.config_path is not None:
+      pex_args.extend(["--config", self.config_path])
+    pex_args.extend(files)
+    return tuple(pex_args)
+
+  def create_execute_request(
+    self,
+    *,
```

Comment:
I'm assuming if we had a way to avoid this we would have, but it makes me sad.

Reply:
Which part?

Reply:
I've been DMing about this. Tansy hadn't seen this Py3-only feature before and confused it with var args like `*args`, which (rightfully so) she finds confusing (as do I).